### PR TITLE
LibELF: Implement dlclose similar to how it is done in glibc

### PR DESCRIPTION
**Pull request by DanShaders**
This resolves multiple fixmes in a way original authors intended them to be resolved. I'm still not convinced we need "proper" dlclose but noone seemed to have any strong opinions about it when I asked on Discord.

CC @supercomputer7, you might want to say something about the design of dlclose promise.

PS: Most of the diff is tests, the actual change is about +120 net added loc.

### CODE REVIEW

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ namespace {
 ByteString s_main_program_path;
 
 // The order of objects here corresponds to the "load order" from POSIX specification.
-OrderedHashMap<ByteString, NonnullRefPtr<ELF::DynamicObject>> s_global_objects;
+OrderedHashMap<ByteString, NonnullOwnPtr<ELF::DynamicObject>> s_global_objects;
```
Comment (ADKaster) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:48]: Not sure bout this one. If we need fancy ref-coutning for Dynamic Object, the proper way would be to implement that custom ref-counting on the object directly, rather than inheriting from RefCounted. RefPtr doesn't require you to use RefCounted, simply that the object has the required ref/unref calls.

File: Userland/Libraries/LibELF/DynamicObject.h [hunk 2]
```diff
@@ -347,6 +347,8 @@ class DynamicObject : public RefCounted<DynamicObject> {
 
     unsigned m_symbol_count { 0 };
 
+    size_t m_dependency_index;
```
Comment (ADKaster) [inline Userland/Libraries/LibELF/DynamicObject.h:384]: `size_t m_dependency_index { 0 };` :)

File: Userland/Libraries/LibELF/DynamicLinker.cpp (hunk 1 above)
Comment (DanShaders) [reply Userland/Libraries/LibELF/DynamicLinker.cpp:48]: I found it easier to reason about the DynamicObjects' lifetimes when there's a single owner and two separate refcounts that aren't obligatory.

File: Userland/Libraries/LibELF/DynamicObject.h (hunk 2 above)
Comment (DanShaders) [reply Userland/Libraries/LibELF/DynamicObject.h:384]: But it is set in constructor...

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 3]
```diff
@@ -39,6 +40,39 @@
...
+
+class ConsecutiveIDAllocator {
+public:
+    size_t allocate()
+    {
+        auto position = m_allocated_ids.find_first<false>();
+        if (position.has_value()) {
+            m_allocated_ids.set(position.value(), true);
+            return position.value();
+        } else {
```
Comment (alimpfard) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:59]: please no.

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 4]
```diff
@@ -45,7 +45,7 @@ namespace {
 ByteString s_main_program_path;
```
Comment (alimpfard) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:46]: I still don't understand why we need this.
Most libcs I know of don't go through all this to support something basically no one relies on.

File: Userland/Libraries/LibELF/DynamicLinker.cpp (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibELF/DynamicLinker.cpp:48]: Looking into the next commit a bit, I'm not sure what the difference between your proposed whacky ownership model and the "normal" RefPtr model; `dlopen_ref` and `dlopen_unref` are just `ref` and `unref`, your manual `unref` call is exactly equivalent to setting the pointer to null. your manual `ref` calls can stay as they are.

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 5]
```diff
@@ -457,6 +457,31 @@ static ErrorOr<void, DlErrorMessage> link_main_library(int flags, DependencyOrde
     return {};
 }
```
Comment (alimpfard) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:463]: > While dlclose is not used often in the wild (and we certainly don't have references to it in Serenity) and is not required by POSIX to do anything, I would rather copy glibc's behavior instead of leaving musl-like no-op behavior.

Why tho?
I'm sorry, but I read that as:
> While dlclose doing anything is literally useless, I want to make it do something.

**Review by alimpfard — COMMENTED** [review]
TLDR: I'm not convinced any of this is necessary, useful, or even good.

**Comment by DanShaders** [comment]
(Off: and that was me being "excessively" emotional with a single "no, no, no" in the review of 21101 :caret:)

Well, I wrote the code, it hasn't felt like something horrible, I thought "fine, let's pr it, I don't care if it gets accepted or not". I think a single objection is enough though to not merge this.

**Comment by alimpfard** [comment]
I'm not really objecting to it, rather asking you what its purpose is -- is there something you want to do that needs this? some code that does something fun when this behaviour is supported? etc etc.
